util/MultiplayerCommon.h

```diff
@@ -20,6 +21,132 @@ FO_COMMON_API extern const std::string SP_SAVE_FILE_EXTENSION;
 
 FO_COMMON_API extern const int ALL_EMPIRES;
 
+class GameRules;
+
+/////////////////////////////////////////////
+// Free Functions
+/////////////////////////////////////////////
+typedef void (*GameRulesFn)(GameRules&); ///< the function signature for functions that add Rules to the GameRules (void (GameRules&))
+
+/** adds \a function to a vector of pointers to functions that add Rules to
+  * the GameRules.  This function returns a boolean so that it can be used to
+  * declare a dummy static variable that causes \a function to be registered as
+  * a side effect (e.g. at file scope:
+  * "bool unused_bool = RegisterGameRules(&foo)"). */
+FO_COMMON_API bool RegisterGameRules(GameRulesFn function);
+
+/** returns the single instance of the GameRules class */
+FO_COMMON_API GameRules& GetGameRules();
+
+/** Database of values that control how the game mechanics function. */
+class FO_COMMON_API GameRules {
+public:
+    enum RuleType : int {
```

Comment:
to elaborate, `Rule` is derived from `OptionsDB::Option` which internally uses `boost::any` to store the values. Options are all defined / registered in C++ code, so the type of the options are all known in the same C++ that gets their value as needed by calling `OptionsDB::Get<T>`, but `Rule` can also be created from definitions in script files, and thus can store one of several types, with no (other) way to know what the type is by inspecting a `Rule`. The enum is thus used to track what type is stored in each `Rule` so it's known what type to `Get`. This is useful within C++ code that otherwise wouldn't know the type stored in a `Rule`, and within the Python API to provide some type-checks when rule values are requested. It's not actually used in the FOCS API, which just does try-catch around the `Get` calls, though in principle could be as well.